traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -0,0 +1,485 @@
+	topologies := map[string]*tc.Topology{}
+	indices := map[int]int{}
+	for index := 0; rows.Next(); index++ {
+		var (
+			name, description string
+			lastUpdated       tc.TimeNoMod
+		)
+		topologyNode := tc.TopologyNode{}
+		topologyNode.Parents = []int{}
+		var parents pq.Int64Array
+		if err = rows.Scan(
+			&name,
+			&description,
+			&lastUpdated,
+			&topologyNode.Id,
+			&topologyNode.Cachegroup,
+			&parents,
+		); err != nil {
+			return nil, nil, errors.New("topology read: scanning: " + err.Error()), http.StatusInternalServerError
+		}
+		for _, id := range parents {
+			topologyNode.Parents = append(topologyNode.Parents, int(id))
+		}
+		indices[topologyNode.Id] = index
+		if _, exists := topologies[name]; !exists {
+			topology := tc.Topology{Nodes: []tc.TopologyNode{}}
+			topologies[name] = &topology
+			topology.Name = name
+			topology.Description = description
+			topology.LastUpdated = &lastUpdated
+		}
+		topologies[name].Nodes = append(topologies[name].Nodes, topologyNode)
+	}
+	if err = rows.Close(); err != nil {
+		userErr, sysErr, errCode := api.ParseDBError(err)
+		return nil, userErr, sysErr, errCode
+	}
+
+	for _, topology := range topologies {
+		nodeMap := map[int]int{}
+		for index, node := range topology.Nodes {
+			nodeMap[node.Id] = index
+		}
+		for _, node := range topology.Nodes {
+			for parentIndex := 0; parentIndex < len(node.Parents); parentIndex++ {
+				node.Parents[parentIndex] = nodeMap[node.Parents[parentIndex]]
+			}
+		}
+		interfaces = append(interfaces, *topology)
+	}
+	return interfaces, nil, nil, http.StatusOK
+}
+
+func (topology *TOTopology) removeParents() error {
+	_, err := topology.ReqInfo.Tx.Exec(deleteParentsQuery(), topology.Name)
+	if err != nil {
+		return errors.New("topology update: error deleting old parents: " + err.Error())
+	}
+	return nil
+}
+
+func (topology *TOTopology) removeNodes(cachegroups *[]string) error {
+	_, err := topology.ReqInfo.Tx.Exec(deleteNodesQuery(), topology.Name, pq.Array(*cachegroups))
+	if err != nil {
+		return errors.New("topology update: error removing old unused nodes: " + err.Error())
+	}
+	return nil
+}
+
+func (topology *TOTopology) addNodes() (error, error, int) {
+	var cachegroupsToInsert []string
+	var indices = make([]int, 0)
+	for index, node := range topology.Nodes {
+		if node.Id == 0 {
+			cachegroupsToInsert = append(cachegroupsToInsert, node.Cachegroup)
+			indices = append(indices, index)
+		}
+	}
+	if len(cachegroupsToInsert) == 0 {
+		return nil, nil, http.StatusOK
+	}
+	rows, err := topology.ReqInfo.Tx.Query(nodeInsertQuery(), topology.Name, pq.Array(cachegroupsToInsert))
+	if err != nil {
+		return nil, errors.New("error adding nodes: " + err.Error()), http.StatusInternalServerError
+	}
+	for _, index := range indices {
+		rows.Next()
+		err = rows.Scan(&topology.Nodes[index].Id, &topology.Name, &topology.Nodes[index].Cachegroup)
+		if err != nil {
+			return api.ParseDBError(err)
+		}
+	}
+	if err = rows.Close(); err != nil {
```

Comment:
closing the rows needs to be in a `defer` in order to ensure that the rows are properly closed however this method may exit

Reply:
Deferring in 73fb323465.